ethereumj-core/src/main/java/org/ethereum/mine/EthashValidationHelper.java

```diff
@@ -0,0 +1,277 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ethereumJ library. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.mine;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.ethereum.config.Constants;
+import org.ethereum.core.BlockHeader;
+import org.ethereum.crypto.HashUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Maintains datasets of {@link EthashAlgo} for verification purposes.
+ *
+ * <p>
+ *     Takes a burden of light dataset caching and provides convenient interface for keeping this cache up to date.
+ *     Featured with full dataset lookup if such dataset is available (created for mining purposes),
+ *     full dataset usage increases verification speed dramatically.
+ *
+ * <p>
+ *     Entry point is {@link #ethashWorkFor(BlockHeader, byte[], boolean)}
+ *
+ * <p>
+ *     Cache management interface: {@link #preCache(long)}, {@link CacheOrder}
+ *
+ * @author Mikhail Kalinin
+ * @since 20.06.2018
+ */
+public class EthashValidationHelper {
+
+    private static final int MAX_CACHED_EPOCHS = 2;
+
+    private static final Logger logger = LoggerFactory.getLogger("ethash");
+
+    public enum CacheOrder {
+        direct,     /** cache is updated to fit main import process, toward big numbers */
+        reverse     /** for maintaining reverse header validation, cache is updated to fit block validation with decreasing numbers */
+    }
+
+    List<Cache> caches = new CopyOnWriteArrayList<>();
+    EthashAlgo ethashAlgo = new EthashAlgo(Ethash.ethashParams);
+    long lastCachedEpoch = -1;
+
+    private CacheStrategy cacheStrategy;
+
+    private static ExecutorService executor;
+
+    public EthashValidationHelper(CacheOrder cacheOrder) {
+        this.cacheStrategy = createCacheStrategy(cacheOrder);
+
+        if (executor == null)
+            executor = Executors.newSingleThreadExecutor((r) -> {
+                Thread t = Executors.defaultThreadFactory().newThread(r);
+                t.setName("ethash-validation-helper");
+                t.setDaemon(true);
+                return t;
+            });
+    }
+
+    /**
+     * Calculates ethash results for particular block and nonce.
+     *
+     * @param cachedOnly flag that defined behavior of method when dataset has not been cached:
+     *                   if set to true  - returns null immediately
+     *                   if set to false - generates dataset for block epoch and then runs calculations on it
+     */
+    public Pair<byte[], byte[]> ethashWorkFor(BlockHeader header, byte[] nonce, boolean cachedOnly) throws Exception {
```

Comment:
I would revisit synchronization on two public methods here, since they could be called concurrently